### PR TITLE
Remove redundant pub(crate) methods for getting resources with id

### LINKED ISSUE #6294: Redundant pub(crate) methods for getting resources with id
- `World::get_non_send_with_id` is redundant, we should remove it and use the public dynamic API `get_resource_by_id`
- `World::get_resource_with_id` is unsound since it has no `Send`/`Sync` checks and is redundant, we should remove it and use the public dynamic API `get_resource_by_id`

### PR DESCRIPTION
**Pull request by targrub**
# Objective

- "Fixes #6294".

## Solution

- Removed `World::get_resource_with_id` and `World::get_non_send_with_id`.
- They were only called in `WorldCell`.  Instead called `World::get_resource_by_id` there and `deref()`.

### CODE REVIEW

File: crates/bevy_ecs/src/world/mod.rs [hunk 1]
```diff
@@ -907,8 +907,10 @@ impl World {
     #[inline]
     pub fn get_resource<R: Resource>(&self) -> Option<&R> {
         let component_id = self.components.get_resource_id(TypeId::of::<R>())?;
+        let p = self.get_resource_by_id(component_id)?;
```
Comment (james7132) [inline crates/bevy_ecs/src/world/mod.rs:910]: This is functionally an equivalent call, but `get_resource_by_id` incurs a Send access check, which is redundant given the constraints on the `Resource` trait. This is a potential performance regression.

File: crates/bevy_ecs/src/world/mod.rs [hunk 2]
```diff
@@ -907,8 +907,10 @@ impl World {
...
         // SAFETY: unique world access
```
Comment (james7132) [inline crates/bevy_ecs/src/world/mod.rs:911]: Not sure how this came to be, but this function does **not**  have unique world access. This safety comment has been incorrect.

File: crates/bevy_ecs/src/world/world_cell.rs [hunk 3]
```diff
@@ -185,9 +185,10 @@ impl<'w> WorldCell<'w> {
         let component_id = self.world.components.get_resource_id(TypeId::of::<T>())?;
         let resource_archetype = self.world.archetypes.resource();
         let archetype_component_id = resource_archetype.get_archetype_component_id(component_id)?;
+        let p = self.world.get_resource_by_id(component_id)?;
```
Comment (james7132) [inline crates/bevy_ecs/src/world/world_cell.rs:188]: Ditto here regarding perf concerns.

**Review by james7132 — CHANGES_REQUESTED** [review]

File: crates/bevy_ecs/src/world/mod.rs (hunk 2 above)
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/world/mod.rs:911]: Pulled out into #6307. Not this PR's problem though.

File: crates/bevy_ecs/src/world/mod.rs (hunk 1 above)
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/world/mod.rs:910]: While this is true, the old `get_resource_with_id` method did not have appropriate thread safety: it didn't specify this requirement, nor did it have a check.

File: crates/bevy_ecs/src/world/world_cell.rs (hunk 3 above)
Comment (alice-i-cecile) [reply crates/bevy_ecs/src/world/world_cell.rs:188]: IMO `WorldCell` should work on `NonSend` resources as well. I'm also way more okay with perf costs here; it's rarer, and deliberately a slower, ref-counted solution to borrow checking woes.

Comment (james7132) [reply crates/bevy_ecs/src/world/world_cell.rs:188]: It does, but it also has dedicated APIs for this, just like World, where the access checks make sense. Here it's always going to evaluate to false and just incur an unnecessary check.

**Comment by alice-i-cecile** [comment]
Moved to #6369.

**Pull request closed without merging**